feat: adds option to OIDC client to require reauthentication

MorrisMorrison opened the pull request:
Hi everyone,
this PR adds an option for OIDC Clients to enforce reauthentication.

Changes:
- adds db migrations for postgres and sqlite to add requires_reauthentication field to oidc_clients
- updates oidc dtos and models with RequiresReauthentication field
- updates oidc_service.HasAuthorizedClient to evaluate if client always requires reauthentication
- updates authorization page to check and enforce reauthentication
- adds form controls in OIDC client settings
- adds translations
- adds e2e test to verify reauthentication is triggered and the user can succesfully sign in afterwards

Fixes: #544


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- New Features
  - Per‑client “Requires Re‑Authentication” option for OIDC.
  - Authorization flow supports WebAuthn reauthentication with token handling.
  - New endpoint to perform passkey reauthentication.
- UI/Style
  - Admin OIDC client form and details show the new setting; widened labels for readability.
- Localization
  - Added English and German strings for the new setting and its description.
- Tests
  - Added end‑to‑end test verifying enforced reauthentication.
- Chores
  - Database migrations for reauthentication flag and tokens; scheduled cleanup of expired tokens.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

kmendell commented:
once ready, feel free to mark as ready for review and ill take a look at it.

kmendell commented:
This looks good to me overall, I would like to wait till elias gets back before approving/merging though. To make sure he agrees as well.

MorrisMorrison commented:
Added the requested changes. Please have a look.

kmendell commented on backend/internal/service/oidc_service.go:
```diff
@@ -187,7 +187,22 @@ func (s *OidcService) Authorize(ctx context.Context, input dto.AuthorizeOidcClie
+		return false, err
+	}
+
+	if client.RequiresReauthentication {
```
This shows the scopes again, im not sure if this was you intended behavior or not, but id think we just want the users to reauthenciate with the passkey again not completely reauthorize the scopes of the application?

ItalyPaleAle replied:
Indeed. I don't know if full re-auth with passkey is required, or just forcing the user to press a button? That would still solve the problem of enforcing user interaction. WDYT?

kmendell commented on frontend/src/routes/authorize/+page.svelte:
```diff
@@ -38,8 +38,8 @@
-			// Get access token if not signed in
-			if (!$userStore?.id) {
+			// Get access token if not signed in OR if client requires re-authentication
+			if (!$userStore?.id || client?.requiresReauthentication) {
```
With my previous comment this would then do the intended behavior, Prompt the user to authenticate with the passkey again, but not have to reaccept the scopes of the application. 

If i understood the issue right, we just want the use to explicitly click the sign in button again, we should still only prompt for a passkey once.

bluewalk replied:
The ticket states:
> Additionally, given the relative ease of authenticating with a passkey, perhaps the "explicit" variant should not only require user interaction, but also re-authentication with the passkey (perhaps 3rd state of "explicit-with-auth").

I'd say make sure you have to re-auth with the passkey again as clicking a simple button to consent doesn't add value/security.

MorrisMorrison replied:
agree with bluewalk, repo owner also mentioned explicitly having to re-auth with passkey in the linked issue. just lmk your decision @kmendell and I will have a look again.

kmendell replied:
Yeah you should have ro reauthenticate once, but right now it prompts you to use the passkey twice, once before and after re-accetping the scopes. removing authoirzation to the client is redundant, we should just force the user to reauth with a passkey, but not make them re accept the scopes of everything. If that makes any sense, my brain is not awake quite yet today lol

MorrisMorrison replied:
yes having to accept the scopes again wasnt intended tbh. just wanted to know if user should only press sign in again or re-auth with passkey. thanks for checking, will update it.

ItalyPaleAle replied:
I think re-accepting the scopes is confusing. All other IdPs I'm familiar with only show the scopes again if they have changed. Showing it every time would make it hard to know if they have changed

ItalyPaleAle commented on backend/internal/dto/oidc_dto.go:
```diff
@@ -31,6 +33,7 @@ type OidcClientCreateDto struct {
 	LogoutCallbackURLs []string                 `json:"logoutCallbackURLs"`
 	IsPublic           bool                     `json:"isPublic"`
 	PkceEnabled        bool                     `json:"pkceEnabled"`
+  RequiresReauthentication bool               `json:"requiresReauthentication"`
```
Please run gofmt (you can also configure your IDE to format Go files automatically on save)

kmendell commented:
Ill try to take a look today!

stonith404 reviewed (changes requested):
Thanks for your PR!

As @ItalyPaleAle has stated currently the re-authentication only happens on client side. The backend would have to make sure that the user has really re-authenticated itself. 

One solution would be create an re-authentication endpoint which returns a very short-lived token similar to then authentication endpoint. The short-lived token then has to be sent with the client authorization request.

stonith404 commented on backend/internal/model/oidc.go:
```diff
@@ -146,3 +147,15 @@ type OidcDeviceCode struct {
+	Base
+	Token     string
+	ExpiresAt datatype.DateTime
+	Used      bool
```
I would prefer to delete the token instead of setting the `Used` flag.

stonith404 commented on backend/internal/service/oidc_service.go:
```diff
@@ -1757,3 +1768,42 @@ func (s *OidcService) IsClientAccessibleToUser(ctx context.Context, clientID str
 	return s.IsUserGroupAllowedToAuthorize(user, client), nil
 }
+
+func (s *OidcService) CreateReauthenticationToken(ctx context.Context, userID string, clientID string) (string, error) {
```
Also same as with the controller here, I would put  `CreateReauthenticationToken` and 
`ValidateReauthenticationToken` function into the `webauth_service.go`.

stonith404 replied:
This function doesn't check if a passkey was used for re-authentication. An attacker could easily bypass the re-authentication by calling this endpoint with curl and receive a re-authentication token.

You would need to solve it similarly to the `VerifyLogin` function.

stonith404 commented on backend/resources/migrations/postgres/20250810012030_oidc_reauthentication_tokens.up.sql:
```diff
@@ -0,0 +1,12 @@
+    updated_at TIMESTAMPTZ DEFAULT CURRENT_TIMESTAMP,
+    token TEXT NOT NULL UNIQUE,
+    expires_at TIMESTAMPTZ NOT NULL,
+    used BOOLEAN NOT NULL DEFAULT FALSE,
```
`used` can be removed, see my comment above. Additionally, please change the timestamp to the current date because we created a new release after the date of this timestamp, which means that the migration doesn't get applied.

stonith404 commented on backend/resources/migrations/postgres/20250710011023_requires_reauthentication_column.up.sql:
```diff
@@ -0,0 +1 @@
+ALTER TABLE oidc_clients ADD COLUMN requires_reauthentication BOOLEAN NOT NULL DEFAULT FALSE;
```
You can merge the `requires_reauthentication_column` and `oidc_reauthentication_tokens` migrations into a single migration.

stonith404 reviewed (changes requested):
Thanks for your changes, I've left some comments.

MorrisMorrison commented:
Hi everyone, thanks for the review!
I added the suggested changes, please have a look.

- moved the reauth endpoint to /webauthn
- moved the logic to webauthn_service
- removed the obsolete RequiresReauthentication flag from oidc_dto
- we directly delete the reauthentication token when consumed, instead of handling it via `used` flag
- renamed `OIDCReauthenticationToken` to just `ReauthenticationToken`
- merge db migrations into one
- updated timestamps of db migrations
- removed obsolete fk on users in migrations
- extended the reauthentication logic to verify passkey was used for reauth, similar to `VerifyLogin`

MorrisMorrison commented:
@stonith404 Hi, I didnt add all the suggestions yet. marked the ones that are done as resolved. will let you know once ready

stonith404 commented:
@MorrisMorrison Oh alright, sorry. Take your time.

stonith404 reviewed (approved):
I had to adapt the re-authentication logic, because I've fixed a potential vulnerability on the main branch in this commit fe003b927ce7772692439992860c804de89ce424.

Looks good to me now, thanks!

kmendell reviewed (approved):
Looks Good to me as well :)

MorrisMorrison commented:
ok great, thanks guys !

ItalyPaleAle commented on backend/internal/controller/webauthn_controller.go:
```diff
@@ -25,6 +25,8 @@ func NewWebauthnController(group *gin.RouterGroup, authMiddleware *middleware.Au
 
 	group.POST("/webauthn/logout", authMiddleware.WithAdminNotRequired().Add(), wc.logoutHandler)
 
+	group.POST("/webauthn/reauthenticate", authMiddleware.WithAdminNotRequired().Add(), rateLimitMiddleware.Add(rate.Every(10*time.Second), 5), wc.reauthenticateHandler)
```
Shouldn't this re-use the same instance of the rate limit middleware used by the other routes? Otherwise, each instance would maintain its own rate-limit, which both allows more requests than what oen may expect, and also will use a lot more memory

stonith404 replied:
For most of the routes we are using the global rate limiter but for a few critical endpoints we create a new instance with a custom limit. IMO it makes sense to keep the rate limit independent for the routes where we set a custom limit because else the limit could get unpredictable.

ItalyPaleAle commented on backend/resources/migrations/postgres/20250814121300_requires_reauthentication.up.sql:
```diff
@@ -0,0 +1,11 @@
+
+CREATE TABLE reauthentication_tokens (
+    id TEXT PRIMARY KEY,
+    created_at TIMESTAMPTZ NOT NULL DEFAULT CURRENT_TIMESTAMP,
```
Do we need to store the time zone? The common practice is to always store times in the DB in UTC, and then format in the appropriate time zone when displaying

ItalyPaleAle reviewed (commented):
A couple of non-blocking comments from me!
